test/OpenFeature.Tests/OpenFeatureHookTests.cs

```diff
@@ -151,6 +164,90 @@ await client.GetBooleanValueAsync("test", false, EvaluationContext.Empty,
             _ = hook2.Received(1).BeforeAsync(Arg.Is<HookContext<bool>>(a => a.EvaluationContext.GetValue("test").AsString == "test"), Arg.Any<ImmutableDictionary<string, object>>());
         }
 
+        [Fact]
+        [Specification("4.1.5", "The `hook data` MUST be mutable.")]
+        public async Task HookData_Must_Be_Mutable()
+        {
+            var hook = Substitute.For<Hook>();
+
+            hook.BeforeAsync(Arg.Any<HookContext<bool>>(), Arg.Any<ImmutableDictionary<string, object>>())
+                .Returns(EvaluationContext.Empty).AndDoes(info =>
+                {
+                    info.Arg<HookContext<bool>>().Data.Set("test-a", true);
+                });
+            hook.AfterAsync(Arg.Any<HookContext<bool>>(), Arg.Any<FlagEvaluationDetails<bool>>(),
+                Arg.Any<ImmutableDictionary<string, object>>()).Returns(new ValueTask()).AndDoes(info =>
+            {
+                info.Arg<HookContext<bool>>().Data.Set("test-b", "test-value");
+            });
+
+            await Api.Instance.SetProviderAsync(new NoOpFeatureProvider());
+            var client = Api.Instance.GetClient("test", "1.0.0");
+
+            await client.GetBooleanValueAsync("test", false, EvaluationContext.Empty,
+                new FlagEvaluationOptions(ImmutableList.Create(hook), ImmutableDictionary<string, object>.Empty));
+
+            _ = hook.Received(1).AfterAsync(Arg.Is<HookContext<bool>>(hookContext =>
+                (bool)hookContext.Data.Get("test-a") == true
+            ), Arg.Any<FlagEvaluationDetails<bool>>(), Arg.Any<ImmutableDictionary<string, object>>());
+            _ = hook.Received(1).FinallyAsync(Arg.Is<HookContext<bool>>(hookContext =>
+                (bool)hookContext.Data.Get("test-a") == true && (string)hookContext.Data.Get("test-b") == "test-value"
+            ), Arg.Any<FlagEvaluationDetails<bool>>(), Arg.Any<ImmutableDictionary<string, object>>());
+        }
+
+        [Fact]
+        [Specification("4.3.2",
+            "`Hook data` **MUST** must be created before the first `stage` invoked in a hook for a specific evaluation and propagated between each `stage` of the hook. The hook data is not shared between different hooks.")]
+        public async Task HookData_Must_Be_Unique_Per_Hook()
+        {
+            var hook1 = Substitute.For<Hook>();
+            var hook2 = Substitute.For<Hook>();
+
+            hook1.BeforeAsync(Arg.Any<HookContext<bool>>(), Arg.Any<ImmutableDictionary<string, object>>())
+                .Returns(EvaluationContext.Empty).AndDoes(info =>
+                {
+                    info.Arg<HookContext<bool>>().Data.Set("hook-1-value-a", true);

```

Comment:
If I'm reading correctly, I think we might be missing an assertion here wherein we'd set some duplicated property (a common key'd prop) on both hook-data instances, and ensure they don't "interfere" with each-other (I think that's implied in 4.3.2).

Please correct me if I'm misunderstanding.

Reply:
Added an assertion using the same name for data in two different hooks with two different values.